File: src/74. Search a 2D Matrix/main.java
// Solved by @kitanoyoru
// https://leetcode.com/problems/search-a-2d-matrix/

class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int rowLen = matrix[0].length;
        int start = 0, mid, end = rowLen - 1, i;

        for (i = 0; i < matrix.length; i++) {
            if (target < matrix[i][0] || target > matrix[i][rowLen - 1]) {
                continue;
            } else {
                while (start <= end) {
                    mid = start + (end - start) / 2;
                    if (matrix[i][mid] == target) {
                        return true;
                    }
                    if (target > matrix[i][mid]) {
                        start = mid + 1;
                    } else {
                        end = mid - 1;
                    }
                }
            }
        }

        return false;
    }
}

File: src/74. Search a 2D Matrix/main.rs
impl Solution {
    pub fn search_matrix(matrix: Vec<Vec<i32>>, target: i32) -> bool {
        let (n, m) = (matrix.len(), matrix[0].len());

        for row in &matrix {
            if target < row[0] || target > row[m - 1] {
                continue;
            } else {
                let (mut start, mut end) = (0, m - 1);
                while start <= end {
                    let mid = start + (end - start) / 2;
                    if row[mid] == target {
                        return true;
                    }

                    if row[mid] < target {
                        start = mid + 1;
                    } else {
                        end = mid - 1;
                    }
                }
            }
        }

        false
    }
}
